Implement stick gate for mouse (for mapping mouse movement to stick movement)

AdmiralCurtiss opened the pull request:
Built on top of #10858, refactored from #10613.

This gets rid of the global state from that PR and makes the part of the implementation in the DInput class a lot more readable. I still want to do some fixups in the actual jail code (and maybe rename it while I'm at it) and I need to figure out if I can get the client area state in there somehow.

Still needs a Linux and Mac impl, too.

AdmiralCurtiss commented:
@Sage-King I've redesigned the configuration for this thing to make it less dependent on window/monitor size. Please check if you can still configure it as you want to use it.

In detail: The 'Sensitivity' option has been split up into Scale and Radius. Scale defines how the outer edge of the window/gate maps to the Cursor inputs, while Radius defines the actual size of the virtual gate. This means that the amount of movement required to move the cursor from one end of the gate to the other is now constant for any given Radius regardless of window/monitor size. I believe this is a better design than what you had before where you effectively sized the gate based on render window height divided by sensitivity.

Actually, after writing this out I think I might drop the Scale options; you could just do the same by adding a multiplication onto the Cursor mappings...

Sage-King commented:
I will definitely check it out, but it might be a few days. Currently trying to get a job and have a lot of interviews scheduled.

Sage-King commented:
- Before this gets pulled, can we add units to the spinners? All of them ideally, but the radius really needs one imo. I thought it was % of the total render window until I checked the tool tip.
- I think having a separate checkbox that enables/disables the mouse centering on start would be good. Could definitely be it's own PR after this gets pulled though
- Could I do a brief write-up on the wiki or website that explains the new feature and how it's intended to be used? Coming back to it with fresh eyes makes me realize that it's a little confusing how to use it. You even had to use two pictures to explain it to devs. The tool tips do a good job explaining it imo, but I'd like somewhere users might check to have an extended explanation.
- The horizontal/vertical scale seems totally functional (even more functional than mine), but the words chosen confused me a little. I've seen this feature called horizontal or x-axis/vertical or y-axis sensitivity in other games. Calling it horizontal/vertical sensitivity seems like a better choice to me. If scale is more commonly used verbiage, I'd be okay with it staying the way it is.
- For convenience, it would be nice to be able to lock/scale them together so it's easier to use it the way I had before. Certainly could be it's own PR after this gets pulled though.
- I'd add some newlines in the tooltips. On my 1920x1080 monitor they go all of the way to the edge of my screen with the default config box positions 
- I also wonder if there's maybe a better way to explain the snapping distance. The mouse only snaps when it leaves the octagonal gate. This helps prevent the mouse from getting stuck in one of the notches, but it makes the feature a little more complicated to explain.

- I see you cached the values. That makes sense to me and I'm not trying to argue that shouldn't have been done. I'm just curious if it was for performance reasons or just helped the code make sense to you.

Thanks for taking the time to revise this :> Your version is SO much cleaner to read. I didn't realize quite how spaghetti my code was :P Not bad for a newb, but it definitely wasn't great thinking back to it.

AdmiralCurtiss commented:
> Before this gets pulled, can we add units to the spinners? All of them ideally, but the radius really needs one imo. I thought it was % of the total render window until I checked the tool tip.

I added a 'pixels' unit for radius and snapping.

> I think having a separate checkbox that enables/disables the mouse centering on start would be good. Could definitely be it's own PR after this gets pulled though

Added this.

> Could I do a brief write-up on the wiki or website that explains the new feature and how it's intended to be used? Coming back to it with fresh eyes makes me realize that it's a little confusing how to use it. You even had to use two pictures to explain it to devs. The tool tips do a good job explaining it imo, but I'd like somewhere users might check to have an extended explanation.

Probably not a bad idea. Wiki seems like the right place for this.

> The horizontal/vertical scale seems totally functional (even more functional than mine), but the words chosen confused me a little. I've seen this feature called horizontal or x-axis/vertical or y-axis sensitivity in other games. Calling it horizontal/vertical sensitivity seems like a better choice to me. If scale is more commonly used verbiage, I'd be okay with it staying the way it is.

I changed the label to Sensitivity.

> For convenience, it would be nice to be able to lock/scale the horizontal/vertical scaling together so it's easier to use it the way I had before. Certainly could be it's own PR after this gets pulled though.

I'm not sure how to do this without a bunch of custom code in the widget so I didn't do this.

> I'd add some newlines in the tooltips. On my 1920x1080 monitor they go all of the way to the edge of my screen with the default config box positions

I changed the tooltips to our nice balloon ones.

> I also wonder if there's maybe a better way to explain the snapping distance. The mouse only snaps when it leaves the octagonal gate. This helps prevent the mouse from getting stuck in one of the notches, but it makes the feature a little more complicated to explain.

Honestly I don't fully understand the snapping distance myself, if you can write up a concise explanation feel free.

> I see you cached the config values in the octagonal mouse gate. That makes sense to me and I'm not trying to argue that shouldn't have been done. I'm just curious if it was for performance reasons or just helped the code make sense to you.

No particular reason, the code already used member variables to hold its data so I just left it as-is, but changed it so that the user doesn't have to explicitly 'initialize' the class when the config/window changed. The actual calculation of those members isn't very expensive so you could probably do it just as well on every call and have the gate fully stateless (it already acts as if it was stateless, anyway).

AdmiralCurtiss commented:
Can I bother someone (@Dentomologist again maybe?) for a macOS implementation here?

Sage-King commented:
>Honestly I don't fully understand the snapping distance myself, if you can write up a concise explanation feel free.

To be honest, my little explanation there is really all there is to know. It only snaps to the notches if the cursor leaves the octagon. 

>Probably not a bad idea. Wiki seems like the right place for this.

Okay, I'll get to that when I have the time after this PR gets pulled.

Apologies for the late response. The git notifications got deleted by accident and I only had a vague recollection of seeing some comments on my PRs by the time I thought about it.
